### PR TITLE
Refactor proxy handling and improve monitor messages

### PR DESCRIPTION
**Pull request by joestringer**
Suggested review commit-by-commit.

Intended functional changes:
* Improved output of trace messages in cases where there were no trace messages before, eg when delivering a packet to the stack from the bpf_hostdev_ingress program, or when unknown protos short-circuit packet handling logic leading to delivery to the stack.
* Avoid always printing the "to proxy port" messages. Use debug messages instead.
* Split bpf_netdev programs into clear "from_netdev" vs. "from_host" to apply at TC ingress / TC egress.
* Defer the functional preparation of the skb to deliver to proxy so it doesn't happen at the original BPF program (eg bpf_netdev attached at cilium_host) but is instead prepared in the bpf_hostdev_ingress program.

These changes are intermediate refactors to simplify the upcoming BPF tproxy PR, also to help bisect any potential regressions.

Related: #9921

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Also seeing lots of messages like this now which aren't overly informative, I wonder if we should avoid logging them:

```
-> 0: 10.56.1.169:53554 -> 172.217.11.170:443 tcp ACK
-> 0: 10.56.1.169:53554 -> 172.217.11.170:443 tcp ACK
-> 0: 10.56.1.169:53554 -> 172.217.11.170:443 tcp ACK
-> 0: 10.56.1.169:53554 -> 172.217.11.170:443 tcp ACK
```

(Just noticed while testing; don't know for sure it's this PR yet but I will confirm before setting this for further review)

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit #10763 .

Otherwise should be ready for review.

**Comment by joestringer** [comment]
@pchaigno this will conflict with your host firewall PR, so let's co-ordinate on how to rebase. This one is a lot smaller so it may make sense for me to just rebase on top of your changes. The logic here is using `at_tc_ingress` rather than `from_host` as the boolean for where the traffic comes from, which unfortunately have the opposite values. I'm inclined to swap over to using the `from_host` nomenclature as that seems like a tidier name.

**Comment by joestringer** [comment]
test-me-please

File: bpf/lib/proxy.h [hunk 1]
```diff
@@ -56,6 +56,23 @@ ctx_redirect_to_proxy(struct __ctx_buff *ctx, __be16 proxy_port)
...
+static __always_inline int
+ctx_redirect_to_proxy_first(struct __ctx_buff *ctx, __be16 proxy_port)
+{
+	int ret = CTX_ACT_OK;
+
+	cilium_dbg(ctx, DBG_CAPTURE_PROXY_POST, proxy_port, 0);
+	ctx->mark = MARK_MAGIC_TO_PROXY | (proxy_port << 16);
+	ctx_change_type(ctx, PACKET_HOST);
+
+	return ret;
```
Comment (qmonnet) [inline bpf/lib/proxy.h]:
```suggestion
	return CTX_ACT_OK;
```
And let's get rid of `ret`? Or why not even returning `void`?

File: bpf/lib/proxy.h [hunk 2]
```diff
@@ -53,7 +56,7 @@ ctx_redirect_to_proxy(struct __ctx_buff *ctx, __be16 proxy_port)
 	cilium_dbg(ctx, DBG_CAPTURE_PROXY_PRE, proxy_port, 0);
 
 	ctx_change_type(ctx, PACKET_HOST); // Required for ingress packets from overlay
-	return CTX_ACT_OK;
+	return result;
```
Comment (qmonnet) [inline bpf/lib/proxy.h]: No need to change this chunk or to add `result` in this commit?

**Review by qmonnet — COMMENTED** [review]
Not marked as pending review, but here are a few nits anyway.
Also there's a conflict on bpf/lib/l3.h. Plus I guess the conflicts with Paul's PR to solve somehow.

File: bpf/lib/proxy.h (hunk 1 above)
Comment (joestringer) [reply bpf/lib/proxy.h]: These changes are in context of other changes I'll make. I can always apply this feedback now but I will change it again in the eBPF tproxy PR. That's probably fair though anyway. Will address.

**Comment by joestringer** [comment]
test-me-please

**Review by jrajahalme — APPROVED** [review]

**Comment by joestringer** [comment]
I pinged @pchaigno offline and we agreed that I should swap the `tc_ingress` flag in the current version of this PR over to `from_host`, which will require reversing all of the boolean logic. I'll work on that.

**Comment by joestringer** [comment]
test-me-please

**Review by qmonnet — APPROVED** [review]

**Pull request merged**